nomad/heartbeat.go

```diff
@@ -142,15 +145,37 @@ func (h *nodeHeartbeater) invalidateHeartbeat(id string) {
 	}
 	h.heartbeatTimersLock.Unlock()
 
-	// Do not invalidate the node since we are not the leader. This check avoids
+	// Do not invalidate the node if we are no longer the leader. This check avoids
 	// the race in which leadership is lost but a timer is created on this
 	// server since it was servicing an RPC during a leadership loss.
-	if !h.IsLeader() {
-		h.logger.Debug("ignoring node TTL since this server is not the leader", "node_id", id)
+	if future := h.raft.VerifyLeader(); future.Error() != nil {
```

Comment:
An alternative approach would be refactoring Node.UpdateStatus's implementation into a new method that *skipped forwarding to the leader* and call that instead. If this server has lost leadership it will return [`raft.ErrNotLeader`](https://pkg.go.dev/github.com/hashicorp/raft#pkg-variables) and fail to apply.

The imprecise `IsLeader` check would be a fine way to skip some work in that case. While `VerifyLeader` is _more_ precise, it's impossible to guarantee leadership hasn't been lost between checking it and calling Node.UpdateStatus anyway.